### PR TITLE
Turn reactive datasource pools into synthetic beans

### LINKED ISSUE #10059: Cannot inject reactive Pool/PgPool into gRPC Singleton
**Describe the bug**
A null reference is passed while injecting a reactive `Pool` or `PgPool` into any bean annotated with `@Singleton` if extension `quarkus-grpc` is added. Injection of my own Singleton works fine.

This mostly affects injection into a gRPC service implementation as these services only work if annotated with `@Singleton`. Injection into `@ApplicationScoped` keeps working as expected. Injection fails with `@Dependent` as well. I did not try any other.

Injection of an `@ApplicationScoped` bean into `@Singleton` still works, probably since its a Proxy. My understanding is `PgPool` is supposed to be a `Singleton` itself, so there shouldn't be a reason to not be able to inject it into a Singleton.

Injection of `PgPool` into Singleton/Dependent works fine in a regular REST project without `quarkus-grpc`.

Tried the following injection methods:
1. package-private field annotated with `@Inject`
2. constructor with `@Inject`
3. constructor without `@Inject`

**Expected behavior**
We should be able to inject reactive Pool/PgPool in gRPC Singletons.

**Actual behavior**
In a gRPC Singleton, injected Pool/PgPool is null at both construction and runtime.

**To Reproduce**
1. Create gRPC service with `quarkus-grpc`, `quarkus-reactive-pg-client`, `quarkus-resteasy`, , `quarkus-resteasy-jackson`, `quarkus-resteasy-mutiny`
2. Add `@Inject Pool pool` to the `@Singleton` extending a gRPC service.
4. In one of the service methods add a Null-Check on `pool` and return an error on null.
3. Launch quarkus
4. Test the method to confirm null.

**Configuration**
```properties
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=test
quarkus.datasource.password=test
quarkus.datasource.reactive.url=postgresql://localhost:5432/test
```

**Environment (please complete the following information):**
  - Output of `uname -a` or `ver`:
```
Darwin REMMACJ5Y8GTFM 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64
```
 - Output of `java -version`:
```
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
```
 - Quarkus version or git rev: 1.5.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
```
------------------------------------------------------------
Gradle 6.5
------------------------------------------------------------

Build time:   2020-06-02 20:46:21 UTC
Revision:     a27f41e4ae5e8a41ab9b19f8dd6d86d7b384dad4

Kotlin:       1.3.72
Groovy:       2.5.11
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          11.0.2 (Oracle Corporation 11.0.2+9)
OS:           Mac OS X 10.15.5 x86_64
```

**Additional context**
Installed features: [cdi, grpc-client, grpc-server, mutiny, reactive-pg-client, resteasy-jackson, resteasy-mutiny, vertx]

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #10059

### CODE REVIEW

**Comment by mkouba** [comment]
CC @yatindra @michalszynkiewicz @tsegismont

**Comment by mkouba** [comment]
Hm, a cycle is detected during build. We need more fixes...

File: extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducer.java [hunk 1]
```diff
@@ -13,43 +14,16 @@
...
      */
     @Singleton
     @Produces
-    public synchronized io.vertx.mutiny.pgclient.PgPool mutinyPgPool() {
-        if (mutinyPgPool == null) {
-            mutinyPgPool = io.vertx.mutiny.pgclient.PgPool.newInstance(pgPool);
-        }
-        return mutinyPgPool;
+    public io.vertx.mutiny.pgclient.PgPool mutinyPgPool() {
+        return io.vertx.mutiny.pgclient.PgPool.newInstance(pgPool);
```
Comment (tsegismont) [inline extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducer.java:26]: Out of curiosity, the bean container will invoke this just once and keep a reference to the result, correct?

**Comment by tsegismont** [comment]
@mkouba LGTM. I'm glad that your update can simplify the producer code by a great margin.

File: extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducer.java (hunk 1 above)
Comment (mkouba) [reply extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducer.java:26]: Yep, it's `@Singleton` so the returned instance is stored in the context.

**Comment by mkouba** [comment]
@tsegismont Ok ;-). But we still have a problem with integration tests. This time it's the `HibernateReactiveDB2Test`...

**Comment by gsmet** [comment]
You need to do the same thing for the MySQL extension :).

**Comment by mkouba** [comment]
> You need to do the same thing for the MySQL extension :).

I did ;-). See the commits.

**Comment by gsmet** [comment]
Busted, I only read the PR title :).

**Comment by mkouba** [comment]
@cescoffier CI is green!

File: extensions/reactive-db2-client/deployment/src/main/java/io/quarkus/reactive/db2/client/deployment/ReactiveDB2ClientProcessor.java [hunk 2]
```diff
@@ -55,11 +58,16 @@ ServiceStartBuildItem build(BuildProducer<FeatureBuildItem> feature,
             return serviceStart;
         }
 
-        RuntimeValue<DB2Pool> db2PoolValue = recorder.configureDB2Pool(vertx.getVertx(), beanContainer.getValue(),
+        RuntimeValue<DB2Pool> db2PoolValue = recorder.configureDB2Pool(vertx.getVertx(),
                 dataSourcesRuntimeConfig, dataSourceReactiveRuntimeConfig, dataSourceReactiveDB2Config,
                 shutdown);
         db2Pool.produce(new DB2PoolBuildItem(db2PoolValue));
 
+        // Synthetic bean for MySQLPool
```
Comment (gsmet) [inline extensions/reactive-db2-client/deployment/src/main/java/io/quarkus/reactive/db2/client/deployment/ReactiveDB2ClientProcessor.java:66]:
```suggestion
        // Synthetic bean for DB2Pool
```

**Review by gsmet — DISMISSED** [review]
I spotted a small typo. Please dismiss when fixed.

**Review by cescoffier — APPROVED** [review]

**Pull request merged**